transport/split/writer.go

```diff
@@ -36,32 +45,99 @@ var _ io.ReaderFrom = (*splitWriterReaderFrom)(nil)
 // A write will end right after byte index prefixBytes - 1, before a write starting at byte index prefixBytes.
 // For example, if you have a write of [0123456789] and prefixBytes = 3, you will get writes [012] and [3456789].
 // If the input writer is a [io.ReaderFrom], the output writer will be too.
-func NewWriter(writer io.Writer, prefixBytes int64) io.Writer {
-	sw := &splitWriter{writer, prefixBytes}
+// It's possible to enable multiple splits with the [AddSplitSequence] option, which adds count splits every skipBytes bytes.
+// Example:
+// prefixBytes = 1, AddSplitSequence(count=2, bytes=6)
+// Array of [0 1 3 2 4 5 6 7 8 9 10 11 12 13 14 15 16 ...] will become
+// [0] [1 2 3 4 5 6] [7 8 9 10 11 12] [13 14 15 16 ...]
+func NewWriter(writer io.Writer, prefixBytes int64, options ...Option) io.Writer {
+	sw := &splitWriter{writer: writer, remainingSplits: []repeatedSplit{}}
+	sw.addSplitSequence(1, prefixBytes)
+	for _, option := range options {
+		option(sw)
+	}
 	if rf, ok := writer.(io.ReaderFrom); ok {
 		return &splitWriterReaderFrom{sw, rf}
 	}
 	return sw
 }
 
+func (w *splitWriter) addSplitSequence(count int, skipBytes int64) {
+	if count == 0 || skipBytes == 0 {
+		return
+	}
+	if w.nextSplitBytes == 0 {
+		w.nextSplitBytes = skipBytes
+		count--
+	}
+	if count > 0 {
+		w.remainingSplits = append(w.remainingSplits, repeatedSplit{count: count, bytes: skipBytes})
+	}
+}
+
+type Option func(w *splitWriter)
+
+// AddSplitSequence will add count splits, each of skipBytes length.
+func AddSplitSequence(count int, skipBytes int64) Option {
+	return func(w *splitWriter) {
+		w.addSplitSequence(count, skipBytes)
+	}
+}
+
 func (w *splitWriterReaderFrom) ReadFrom(source io.Reader) (int64, error) {
-	reader := io.MultiReader(io.LimitReader(source, w.prefixBytes), source)
-	written, err := w.rf.ReadFrom(reader)
-	w.prefixBytes -= written
+	var written int64
+	for w.nextSplitBytes > 0 {
+		expectedBytes := w.nextSplitBytes
+		n, err := w.rf.ReadFrom(io.LimitReader(source, expectedBytes))
+		written += n
+		w.advance(n)
+		if err != nil {
+			return written, err
+		}
```

Comment:
likely need to handle `io.EOF` generated by the `LimitReader`

Reply:
wait, sorry, I think I'm wrong here

Reply:
ReadFrom is not supposed to return EOF. It just stops the loop and returns nil